Sources/ArgumentParser/Utilities/SubcommandDiscovery.swift

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import _Runtime
+@_implementationOnly import Foundation
+
+let subcommandLock = NSLock()
+var discoveredSubcommands: [UnsafeRawPointer: [ParsableCommand.Type]] = [:]
+
+// Use runtime metadata information to find nested subcommands automatically.
+// When getting the subcommands for a ParsableCommand, we look at all types who
+// conform to ParsableCommand within the same module. If we find one who
+// is nested in the base command that we're looking at now, we add it as a
+// subcommand.
+//
+//      struct Base: ParsableCommand {}
+//
+//      extension Base {
+//        // This is considered an automatic subcommand!
+//        struct Sub: ParsableCommand {}
+//      }
+//
+func discoverSubcommands(for type: Any.Type) -> [ParsableCommand.Type] {
+  // Make sure that only classes, structs, and enums are checked for.
+  guard let selfMetadata = metadata(for: type),
+        selfMetadata.kind != .existential else {
+    return []
+  }
+  
+  subcommandLock.lock()
+  
+  defer {
+    subcommandLock.unlock()
+  }
+  
+  guard !discoveredSubcommands.keys.contains(selfMetadata.pointer) else {
+    return discoveredSubcommands[selfMetadata.pointer].unsafelyUnwrapped
```

Comment:
I think we should avoid the use of `unsafelyUnwrapped` unless absolutely necessary for performance. My intuition is that in this kind of code we can get away with using the force unwrap operator instead.